crypto/x509/v3_ncons.c

```diff
@@ -782,50 +783,57 @@ static int nc_email(ASN1_IA5STRING *eml, ASN1_IA5STRING *base)
 static int nc_uri(ASN1_IA5STRING *uri, ASN1_IA5STRING *base)
 {
     const char *baseptr = (char *)base->data;
-    const char *hostptr = (char *)uri->data;
-    const char *p = ia5memchr(uri, (char *)uri->data, ':');
+    char *uri_copy;
+    char *scheme;
+    char *host;
     int hostlen;
+    int ret;
 
-    /* Check for foo:// and skip past it */
-    if (p == NULL
-            || IA5_OFFSET_LEN(uri, p) < 3
-            || p[1] != '/'
-            || p[2] != '/')
-        return X509_V_ERR_UNSUPPORTED_NAME_SYNTAX;
-    hostptr = p + 3;
-
-    /* Determine length of hostname part of URI */
+    if ((uri_copy = OPENSSL_strndup((const char *)uri->data, uri->length)) == NULL)
+        return X509_V_ERR_UNSPECIFIED;
 
-    /* Look for a port indicator as end of hostname first */
+    if (!OSSL_parse_url(uri_copy, &scheme, NULL, &host, NULL, NULL, NULL, NULL, NULL)) {
+        OPENSSL_free(uri_copy);
+        return X509_V_ERR_UNSUPPORTED_NAME_SYNTAX;
+    }
 
-    p = ia5memchr(uri, hostptr, ':');
-    /* Otherwise look for trailing slash */
-    if (p == NULL)
-        p = ia5memchr(uri, hostptr, '/');
+    /* Make sure the scheme is there */
+    if (scheme == NULL) {
```

Comment:
Argh, I was again confused, this time by the `HTTP_parse_url` code. 
So please (also to be on the safe side for any future changes here):
```suggestion
    if (scheme == NULL || *scheme == '\0') {
```

Reply:
I do not understand why this is necessary. In case the URI is incorrect wrt scheme, the function will fail so checking for `NULL` is sufficient. Or put another way, if the function returns success, we're guaranteed that scheme is not an empty string. Please check the implementation of `OSSL_parse_url`.

Reply:
As mentioned, the original code returned failure if no scheme (indicated by `:`) followed by // is present at all,
and I believe we should keep this behavior.
I just double-checked: in such case `OSSL_parse_url()` happily reports an empty string for the scheme, with `scheme` being non-NULL. So we should better catch this on success of `OSSL_parse_url()`.

Reply:
> As mentioned, the original code returned failure if no scheme (indicated by `:`) followed by // is present at all, and I believe we should keep this behavior. 

Ah, I see now, you're talking about the case of `://` missing completely. Yes, you're right, that's something we need to handle, sorry for the confusion. Will amend. :+1: